### ISSUE TITLE
Doc issue: clarification of setting :response in an intercepter

### ISSUE THREAD

**Issue by deg** [pedestal/pedestal-docs#70]
I was recently bitten by lack of knowledge, writing an interceptor for a Vase app.

I needed to set the context's :response and then let it be processed by existing Vase code (e.g., converting the response to transit).

I initially did this in the :enter of my interceptor, and was surprised to discover that the remaining interceptors were not called. (The proximate reason is https://github.com/pedestal/pedestal/blob/master/service/src/io/pedestal/http/impl/servlet_interceptor.clj#L373, and I assume that the intention is deliberate).

It was easy to fix my code; I just needed to set the response in my :leave. But, this was not obvious at first, and would be helped by more documentation and examples of how to write interceptors. (Ideally, also with a link to this part of the Pedestal docs from the Vase docs).

**Comment by ohpauleez** [comment]
Thanks for using Pedestal and Vase, and reporting the issue to help improve the docs and experience!

Please see my response [on the initial issue reported on Pedestal](https://github.com/pedestal/pedestal/issues/511).  Cross posted for conversation purposes:

I believe this is well documented in the [Pedestal docs](http://pedestal.io/reference/interceptors) and covered in various other channels, like my recent [Clojure/West talk on Pedestal and Vase](https://www.youtube.com/watch?v=_Cf-STRvFy8).

In the context, there is a sequence of :terminators, predicate functions that determine if processing is complete or not. The default terminator for all HTTP application is the presence of a :response in the context - Once there is a response, the system should send it immediately. When processing stops because of a terminator, the leave stage is called for all interceptors on the stack (ie: all interceptors that have already been called/processed via :enter).

Your interceptor should still be called on the :enter stage, it just needs to be placed later in the queue (so your other interceptors can also have a chance to perform their processing).
- - - - -

Additionally, if you ever hit a snag or stumble, please feel free to reach out on #pedestal in the Clojurians Slack or on the [Pedestal mailing list](https://groups.google.com/forum/#!forum/pedestal-users).

**Comment by deg** [comment]
I agree that the file you reference (http://pedestal.io/reference/interceptors) is the right place to document this. But, I don't think the issue is addressed enough there. The only relevant line is the final line in the doc: "Because a handler takes one kind of thing (request) and returns a different kind of thing (response), it can only be used in the last position of a stack."  At least for me, this was not clear enough at first.

I apologize that I don't have time to watch your full video now, but will try to view it later.

Re your fuller answer, thanks!  That paragraph, with maybe a bit more elaboration, would be a great addition to the docs. 

Style question: why is it better to call my interceptor on the :enter stage? 
I now (see below) see that this works, but it seems more error-prone. If someone adds another interceptor later, the code will break. Why is it not better stylistically to set :response on :leave? (I understand that it makes no real difference -- the final :enter immediately precedes the initial :leave; this is just a style/robustness question since your answer favors :enter).

TL;DR of the rest of this message - I was confused, and am now trying to understand why.

Note: I did spend the last few minutes trying to prove you incorrect, based on my experience yesterday.
But, the simple case -- changing `:leave` to `:enter` at https://github.com/deg/cljtoys-Receipts/blob/master/server/src/receipts_server/interceptors.clj#L67 does seem to work perfectly.
And, each of my failing cases from yesterday that I still have in my reflog have other problems that I can now understand.

Part of what made the model hard for me to understand yesterday is that Vase (and also Pedestal) do so much implicitly/declaratively, so there were no easy places for me to peek in to refine my mental model.
My model was completely off, because it had not even occurred to me that `(assoc context :response ...)` was a side-effecting operation when performed anywhere but the end of the interceptor chain. From there, I spun off a pile of wrong theories that led to bad tests and wrong conclusions.

The exact details of my confusion don't matter. My real take-away here is that it is inherently confusing that setting `:response` affects what interceptors will run.  I assume this behavior is for very good reasons and won't be changed so this just calls for a bit more documentation.

**Comment by ohpauleez** [comment]
I think that watching the video will clear everything up for you.  It seems like you have the model of processing almost exactly opposite/inverse.  All processing and behavior is explicit, and all of it is set and controlled within the context.  Since interceptors take and return a context, you have full control over processing at all times.

**Why should I use `:enter` instead of `:leave` for responses/actions?**
Processing `:leave` implies one of two things: Either a terminator predicate function returned true (the default case is that there is a `:response` in the context) or you hit the end of the chain without satisfying any terminator.  Using `:leave` to attach a valid response means you've done more processing than needed/required.  The only reason for attaching a `:response` within a `:leave` would be something like a ["Not Found" interceptor](https://github.com/pedestal/pedestal/blob/master/service/src/io/pedestal/http.clj#L92-L100) to signify that nothing could handle/satisfy the request.

Pedestal is an implementation of the Interceptor pattern.  Given a queue interceptors (`[i1 i2 i3 i4 ...]`), Pedestal will call their `:enter` stage in order, attempting to fulfill processing (until a function within `:terminators` returns true).  Put another way, `:enter` is for handling requests/payloads, `:leave` is for cleaning up and post-processing a response.

**Comment by deg** [comment]
OK, I watched your video. It is very well done and gives a great overview of Pedestal and Vase. Thanks!

But:

I don't understand why you think that I "have the  model of processing almost exactly opposite/inverse". After watching the video, I'm more convinced that I do understand the model, and that I've merely identified a surprise in Pedestal's implementation of HTTP server support.

I assume that this surprise is necessary for very valid reasons. I'm asserting only that _it needs to be documented better because it is, indeed a surprise_.

The video did not address my point  of surprise -- that touching `(:response ctx)` cause a change in the interceptor chain.  To me, this is surprising.

I 100% get that (a) there is a chain of interceptors with paired enter/leave functions; (b) that the chain itself is passed to each interceptor as part of the context; and (c) that any interceptor can freely modify the chain in any way.

This is all wonderful, all consistent, and all makes sense.  My surprise came simply from the fact that setting the context's `:response`, as compared to setting its `:foo` or `:bar`, triggers a modification of the interceptor chain.

(I'll save my follow-on question re enter vs leave for later. It is truly secondary, and I'd rather make sure that I'm on the right page for the main issue before I blabber on about a detail).

**Comment by ohpauleez** [comment]
I think we could update the docs to mention this in a few places.  In addition to expanding the conversation about `:terminators` in the Interceptor section, we might more explicitly call out the terminator in the [Servlet Chain Provider doc](http://pedestal.io/reference/servlet-interceptor) and maybe a sentence or two in the [second guide where this topic is explained](http://pedestal.io/guides/hello-world-content-types).

In the future, if you're curious about the terminators for a given chain, you can always inspect the `:terminators` key in the context or see how the chain provider establishes the initial context (as you linked to originally -- I usually just grep around for `:terminators` or `terminate-when`)

**Comment by deg** [comment]
Sounds perfect. You also might want to eventually add links from the docs to the relevant videos or even to moments within videos. (Personally, I prefer docs to video, but for the many folks who are the opposite, this would be great).

This addresses my initial issue, thanks!

Re my second issue (enter vs leave), I just looked at your answer again. Either I missed your last paragraph earlier, or you edited it in after my first quick reading. Either way. yes, I agree with what you say there.

This is an important explanation (which seems obvious to me now, but definitely was not obvious yesterday). This discussion can be fit nicely into the docs in the ["second guide"](http://pedestal.io/guides/hello-world-content-types) you cited, where Mike does exactly this: putting `coerce-to` in a `:leave`.

I've read this doc previously, and completely glossed over the fact that he used :leave. This is a great excuse to expand the doc there and explain this point.

At the level you discussed in your video, :leave vs :enter are not that different. As you said "if data flows in to an interceptor, you call the enter function. if data flows out, you call the leave function". But, an adapter (like `coerce-to` or my cvs function) is more ambiguous, and can be viewed as either consuming or producing data, depending on your perspective. (Similarly, your first reaction to my case was that I should use an :enter, and your later instinct, after seeing the code, was to use a :leave).

A stumbling block here for the new user is that there are two reasons to choose between :enter and :leave for an adapter interceptor, and they may contradict each other:
1) the stylistic and perspective arguments. Here, reasonable people and projects can disagree, or choose based on external reasons.
2) The surprise of the :response behavior that mandates that an adapter *can not* set the response on the :enter chain if there is any chance that any interceptor will ever be added later in the chain.
